### PR TITLE
feat(operation): spectral / condition / rank matrix properties (#244 batch 3)

### PR DESCRIPTION
**Pull request by Ravenwater**
Batch 3 of the properties module (#244): numeric queries backed by the existing
dense **SVD** and **eigenvalue** solvers (both copy `A` internally, so the
caller's matrix is unchanged).

## New API — `namespace mtl` (`operation/spectral_properties.hpp`)

| Function | Backed by | Semantics |
|---|---|---|
| `spectral_radius(A)` | eigenvalue | max \|eigenvalue\| |
| `condition_number(A)` | SVD | 2-norm σ_max/σ_min; `+inf` when rank-deficient |
| `rcond(A)` | SVD | reciprocal σ_min/σ_max ∈ [0, 1] (safe when singular) |
| `numerical_rank(A[, tol])` | SVD | count of singular values above a threshold |
| `nullity(A[, tol])` | SVD | `num_cols − numerical_rank` |

**`numerical_rank` threshold:** default cutoff `max(m,n)·eps·σ_max` (the standard
rank-revealing cutoff); pass `tol` for an explicit absolute threshold on the
singular values.

**Edge cases:** empty (0×0) → `spectral_radius` 0, `condition_number`/`rcond` 1;
rectangular ranks reduce over the `min(m,n)` singular values.

## Tests (`test_spectral_properties.cpp`)

- Spectral radius: diagonal, 2-D rotation (±i), SPD (largest eigenvalue), empty.
- Condition number / rcond: diagonal, identity, **randspd** with prescribed
  eigenvalues (cond = 4), rank-deficient (huge cond / near-zero rcond).
- Rank / nullity: full-rank, rank-deficient (zero singular value), tolerance
  sweep, rectangular full-column-rank.

Checks are written to hold on **both** the LAPACK and the in-house SVD/eigen
paths — the rank-deficient condition-number is tested as "very large or infinite"
(the in-house SVD gives a tiny-but-nonzero σ_min) and rank-deficient counts pass
an explicit `tol`. Verified against a `-DMTL5_WITH_LAPACK=ON` build locally.

## Scope note

Phase 3 of 4 in #244. Deferred: `is_orthogonal`/`is_unitary`/`is_normal` (product-
based, different flavor) and `inertia`/`is_indefinite` — a small follow-up slice.
Next: batch 4 (tensor predicates).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added matrix spectral/conditioning/rank query functions: spectral radius, condition number, reciprocal condition number, numerical rank, and nullity.
  * Added consistent edge-case behavior for empty and rank-deficient matrices, including tolerance-based rank determination.
  * Made these operations available via the main MTL umbrella include.

* **Tests**
  * Added unit tests validating these queries across diagonal, rotation, SPD, rectangular, empty, and rank-deficient matrices.

### CODE REVIEW

File: include/mtl/operation/spectral_properties.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+} // namespace detail
+
+/// Spectral radius: max |eigenvalue|. Requires a square matrix; the empty (0x0)
+/// matrix has spectral radius 0.
+template <Matrix M>
+magnitude_t<typename M::value_type> spectral_radius(const M& A) {
+    using mag_t = magnitude_t<typename M::value_type>;
+    using std::abs;
+    if (A.num_rows() == 0) return mag_t(0);
+    auto eigs = eigenvalue(A);
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/spectral_properties.hpp:51]: Fixed in e2aa8b1: `spectral_radius` now asserts the square precondition (`num_rows() == num_cols()`) before the `num_rows() == 0` fast path, matching `eigenvalue`s contract. A 0x0 matrix is square and returns 0; a 0xN matrix now trips the assertion instead of silently returning 0.

File: include/mtl/operation/spectral_properties.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+// spectral_radius, condition_number, rcond, numerical_rank, nullity.
+//
+// These wrap the existing dense SVD and eigenvalue solvers (both take A by
+// const reference and copy internally), so the caller's matrix is unchanged.
+// Cost is that of the underlying decomposition (SVD / QR-iteration), for dense
+// matrices with a real floating value type.
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/spectral_properties.hpp:14]: Fixed in e2aa8b1: added direct `#include <algorithm>` (and `<cassert>`), and `spectral_radius` now asserts `num_rows() == num_cols()` before the empty-matrix fast path so a 0xN input no longer silently returns 0.

File: include/mtl/operation/spectral_properties.hpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    using mag_t = magnitude_t<typename M::value_type>;
+    using size_type = typename M::size_type;
+    using std::abs;
+    const size_type mn = std::min(A.num_rows(), A.num_cols());
+    std::vector<mag_t> sv;
+    sv.reserve(static_cast<std::size_t>(mn));
+    if (mn == 0) return sv;
+    auto [U, S, V] = svd(A);
+    (void)U; (void)V;
+    for (size_type i = 0; i < mn; ++i) sv.push_back(abs(S(i, i)));
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/mtl/operation/spectral_properties.hpp]: Fixed in e2aa8b1: the `singular_values` helper (and the loop counters in `spectral_radius`/`numerical_rank`) now use `std::size_t` for local dimensions and indices per the portability guideline.
